base/commands/migration/migration_stages.go

```diff
@@ -0,0 +1,328 @@
+				Name: m.Name,
+				Type: m.Type,
+			})
+		}
+	} else {
+		return nil, fmt.Errorf("no datastructures found to migrate")
+	}
+	return dss, nil
+}
+
+func saveMemberLogs(ctx context.Context, ec plug.ExecContext, ci *hazelcast.ClientInternal, migrationID string) error {
+	for _, m := range ci.OrderedMembers() {
+		l, err := ci.Client().GetList(ctx, DebugLogsListPrefix+m.UUID.String())
+		if err != nil {
+			return err
+		}
+		logs, err := l.GetAll(ctx)
+		if err != nil {
+			return err
+		}
+		for _, line := range logs {
+			ec.Logger().Info(fmt.Sprintf("[%s_%s] %s", migrationID, m.UUID.String(), line.(string)))
+		}
+	}
+	return nil
+}
+
+func saveReportToFile(ctx context.Context, ci *hazelcast.ClientInternal, migrationID, fileName string) error {
+	report, err := fetchMigrationReport(ctx, ci, migrationID)
+	if err != nil {
+		return err
+	}
+	if report == "" {
+		return nil
+	}
+	f, err := os.Create(fmt.Sprintf(fileName))
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	return os.WriteFile(fileName, []byte(report), 0600)
+}
+
+func waitForMigrationToBeInProgress(ctx context.Context, ci *hazelcast.ClientInternal, migrationID string) error {
+	for {
+		status, err := fetchMigrationStatus(ctx, ci, migrationID)
+		if err != nil {
+			if errors.Is(err, migrationStatusNotFoundErr) {
+				continue // migration status will not be available for a while, so we should wait for it
+			}
+			return err
+		}
+		if Status(status) == StatusInProgress {
+			return nil
+		}
+	}
+}
+
+type OverallMigrationStatus struct {
+	Status               Status              `json:"status"`
+	Logs                 []string            `json:"logs"`
+	Errors               []string            `json:"errors"`
+	Report               string              `json:"report"`
+	CompletionPercentage float32             `json:"completionPercentage"`
+	Migrations           []DSMigrationStatus `json:"migrations"`
+}
+
+type DataStructureInfo struct {
+	Name string
+	Type string
+}
+
+type DSMigrationStatus struct {
+	Name                 string  `json:"name"`
+	Type                 string  `json:"type"`
+	Status               Status  `json:"status"`
+	CompletionPercentage float32 `json:"completionPercentage"`
+	Error                string  `json:"error"`
+}
+
+func fetchMigrationStatus(ctx context.Context, ci *hazelcast.ClientInternal, migrationID string) (string, error) {
+	q := fmt.Sprintf(`SELECT JSON_QUERY(this, '$.status') FROM %s WHERE __key='%s'`, StatusMapName, migrationID)
+	res, err := ci.Client().SQL().Execute(ctx, q)
+	if err != nil {
+		return "", err
+	}
+	it, err := res.Iterator()
+	if err != nil {
+		return "", err
+	}
+	if it.HasNext() { // single iteration is enough that we are reading single result for a single migration
+		row, err := it.Next()
+		if err != nil {
+			return "", err
+		}
+		r, err := row.Get(0)
+		var m string
+		if err = json.Unmarshal(r.(serialization.JSON), &m); err != nil {
+			return "", err
+		}
+		return m, nil
```

Comment:
Same comment about checking the error from `row.Get` and `serialization.JSON == []byte`.

Reply:
https://github.com/hazelcast/hazelcast-commandline-client/pull/398/commits/0c7028cdae9d4dd0c53fd5a322d35f0b0c3e9fef